cli/lib/output-mapper.ts

```diff
@@ -104,23 +120,51 @@ export async function mapDimensions(this: Pick<Context, 'variables'>) {
     return value
   }
 
-  const { default: map } = await importDynamic('barnard59-base/map.js')
+  // We store non-observation quads in memory and inject them into the stream
+  // after all the observation-specific steps
+  const nonObservationQuads = $rdf.dataset()
+  this.variables.set('nonObservationQuads', nonObservationQuads)
 
-  return map(async (quad: Quad) => {
+  const handleQuad = async (quad: Quad) => {
     const { subject, predicate, object, graph } = quad
     const mappingTerm = getDimensionMapping(predicate)
     if (mappingTerm?.value) {
       if (object.equals(undef)) {
-        return $rdf.quad(subject, predicate, cube.Undefined, graph)
+        return [$rdf.quad(subject, predicate, cube.Undefined, graph)]
       }
 
       const mappedValue = await getMappedValue(mappingTerm.value, object)
       if (mappedValue?.termType === 'NamedNode') {
-        return $rdf.quad(subject, predicate, mappedValue, graph)
+        const predicateOriginal = $rdf.namedNode(predicate.value + '#_original')
+
+        nonObservationQuads.add($rdf.quad(predicateOriginal, rdf.type, cc.OriginalValuePredicate, graph))
+        nonObservationQuads.add($rdf.quad(predicateOriginal, schema.about, predicate, graph))
+
+        return [
+          $rdf.quad(subject, predicate, mappedValue, graph),
+          $rdf.quad(subject, predicateOriginal, object, graph),
+        ]
       }
     }
 
-    return quad
+    return [quad]
+  }
+
+  return flatMap(handleQuad)
+}
+
+export function injectNonObservationQuads(this: Pick<Context, 'variables'>) {
+  const nonObservationQuads = this.variables.get('nonObservationQuads')
```

Comment:
I'm not exactly fond of the name. Maybe make it more specific to the purpose of keeping original values?